compiler/passes/src/option_lowering/ast.rs

```diff
@@ -0,0 +1,707 @@
+            Expression::Err(e) => self.reconstruct_err(e, additional),
+            Expression::Path(e) => self.reconstruct_path(e, additional),
+            Expression::Literal(e) => self.reconstruct_literal(e, additional),
+            Expression::Locator(e) => self.reconstruct_locator(e, additional),
+            Expression::MemberAccess(e) => self.reconstruct_member_access(*e, additional),
+            Expression::Repeat(e) => self.reconstruct_repeat(*e, additional),
+            Expression::Ternary(e) => self.reconstruct_ternary(*e, additional),
+            Expression::Tuple(e) => self.reconstruct_tuple(e, additional),
+            Expression::TupleAccess(e) => self.reconstruct_tuple_access(*e, additional),
+            Expression::Unary(e) => self.reconstruct_unary(*e, additional),
+            Expression::Unit(e) => self.reconstruct_unit(e, additional),
+        };
+
+        // Optionally wrap in an optional if expected type is `Optional<T>`
+        if let Some(Type::Optional(OptionalType { inner })) = additional {
+            let actual_expr_type =
+                self.state.type_table.get(&expr.id()).expect(
+                    "Type table must contain type for this expression ID; IDs are not modified during lowering",
+                );
+
+            if actual_expr_type.can_coerce_to(inner) {
+                return (self.wrap_optional_value(expr, *inner.clone()), stmts);
+            }
+        }
+
+        (expr, stmts)
+    }
+
+    fn reconstruct_array_access(
+        &mut self,
+        mut input: ArrayAccess,
+        _additional: &Self::AdditionalInput,
+    ) -> (Expression, Self::AdditionalOutput) {
+        let (array, mut stmts_array) = self.reconstruct_expression(input.array, &None);
+        let (index, mut stmts_index) = self.reconstruct_expression(input.index, &None);
+
+        input.array = array;
+        input.index = index;
+
+        // Merge side effects
+        stmts_array.append(&mut stmts_index);
+
+        (input.into(), stmts_array)
+    }
+
+    fn reconstruct_associated_function(
+        &mut self,
+        mut input: AssociatedFunctionExpression,
+        _additional: &Option<Type>,
+    ) -> (Expression, Self::AdditionalOutput) {
+        match CoreFunction::from_symbols(input.variant.name, input.name.name) {
+            Some(CoreFunction::OptionalUnwrap) => {
+                let [optional_expr] = &input.arguments[..] else {
+                    panic!("guaranteed by type checking");
+                };
+
+                let (reconstructed_optional_expr, mut stmts) =
+                    self.reconstruct_expression(optional_expr.clone(), &None);
+
+                // Access `.val` and `.is_some` from reconstructed expression
+                let val_access = MemberAccess {
+                    inner: reconstructed_optional_expr.clone(),
+                    name: Identifier::new(Symbol::intern("val"), self.state.node_builder.next_id()),
```

Comment:
Should these symbols have a "unique" tag in them so that they don't clash with identifiers?

Reply:
I'm convinced! All set here.